File: algo-01/art-optimal.py
'''
root@raspberrypi:~/projet-smb215# python3 algo-01/art-optimal.py
root@raspberrypi:~/projet-smb215# perf stat python3 algo-01/art-optimal.py
'''

import os
DIR = os.path.dirname(os.path.abspath(__file__))
INPUT_FILE = DIR + "/input.txt"
OUTPUT_FILE = DIR + "/solution.txt"

with open(INPUT_FILE) as f:
    baseX, baseY = map(int, f.readline().strip().split(','))
    baseGrid = []
    for line in f:
        baseGrid.append(line.strip())
#print(f'BaseX: {baseX}, BaseY: {baseY}')
result = []
for y in range(baseY):
    for x in range(baseX):
        if baseGrid[y][x] == "#":
            result.append(f'FILL,{x},{y},1')
with open(OUTPUT_FILE, 'w') as f:
    f.write('\n'.join(result))
print('Res:', len(result))

File: algo-01/art-optimal.js
/*
root@raspberrypi:~/projet-smb215# node algo-01/art-optimal.js
root@raspberrypi:~/projet-smb215# perf stat node algo-01/art-optimal.js
*/

const fs = require('fs');

INPUT_FILE = __dirname + "/input.txt";
OUTPUT_FILE = __dirname + "/solution.txt";

fs.readFile(INPUT_FILE, 'utf8', (err, data) => {
    if (err) {
        return console.error(`Unable to open the file: ${err}\n`);
    }

    data = data.split('\r\n');
    const [baseX, baseY] = data[0].split(',').map(value => parseInt(value));

    // console.log(`BaseX: ${baseX}, BaseY: ${baseY}`);

    let result = [];
    const baseGrid = data.slice(1);
    for (let y = 0; y < baseY; y++) {
        for (let x = 0; x < baseX; x++) {
            if (baseGrid[y][x] === "#") {
                result.push(`FILL,${x},${y},1`);
            }
        }
    }

    fs.writeFile(OUTPUT_FILE, result.join('\n'), {flag: 'w+'}, (err) => {
        if (err) {
            return console.error(`Unable to write the file: ${err}\n`);
        }
        console.log(`Res: ${result.length}`);
    });
});